systemtest/edu/washington/escience/myria/systemtest/ParallelIngestS3Test.java

```diff
@@ -0,0 +1,149 @@
+    server.submitQueryPlan(new SinkRoot(new EOSSource()), workerPlansParallelIngest).get();
+    assertEquals(2556, server.getDatasetStatus(relationKey).getNumTuples());
+  }
+
+  @Test
+  public void diffParallelIngestTest() throws Exception {
+
+    /* Ingest in parallel */
+    RelationKey relationKeyParallelIngest = RelationKey.of("public", "adhoc", "ingestParallel");
+    Map<Integer, RootOperator[]> workerPlansParallelIngest = new HashMap<Integer, RootOperator[]>();
+    int workerCounterID = 1;
+    for (int workerID : workerIDs) {
+      AmazonS3Source s3Source = new AmazonS3Source(customerTableAddress);
+      CSVFileScanFragment scanFragment =
+          new CSVFileScanFragment(s3Source, customerSchema, workerCounterID, workerIDs.length, ',', null, null, 0);
+      workerPlansParallelIngest.put(workerID, new RootOperator[] { new DbInsert(scanFragment,
+          relationKeyParallelIngest, true) });
+      workerCounterID++;
+    }
+    server.submitQueryPlan(new SinkRoot(new EOSSource()), workerPlansParallelIngest).get();
+    assertEquals(300000, server.getDatasetStatus(relationKeyParallelIngest).getNumTuples());
+
+    /* Ingest the through the coordinator */
+    RelationKey relationKeyCoordinatorIngest = RelationKey.of("public", "adhoc", "ingestCoordinator");
+    server.ingestDataset(relationKeyCoordinatorIngest, server.getAliveWorkers(), null, new FileScan(new UriSource(
+        customerTableAddress), customerSchema, ',', null, null, 0), new RoundRobinPartitionFunction(workerIDs.length));
+    assertEquals(300000, server.getDatasetStatus(relationKeyCoordinatorIngest).getNumTuples());
+
+    /* do the diff at the first worker */
+    final Map<Integer, RootOperator[]> workerPlansDiff = new HashMap<Integer, RootOperator[]>();
+
+    DbQueryScan scanParallelIngest = new DbQueryScan(relationKeyParallelIngest, customerSchema);
+    ExchangePairID receiveParallelIngest = ExchangePairID.newID();
+    CollectProducer sendToWorkerParallelIngest =
+        new CollectProducer(scanParallelIngest, receiveParallelIngest, workerIDs[0]);
+
+    DbQueryScan scanCoordinatorIngest = new DbQueryScan(relationKeyCoordinatorIngest, customerSchema);
+    ExchangePairID receiveCoordinatorIngest = ExchangePairID.newID();
+    CollectProducer sendToWorkerCoordinatorIngest =
+        new CollectProducer(scanCoordinatorIngest, receiveCoordinatorIngest, workerIDs[0]);
+
+    CollectConsumer workerConsumerParallelIngest =
+        new CollectConsumer(customerSchema, receiveParallelIngest, workerIDs);
+    CollectConsumer workerConsumerCoordinatorIngest =
+        new CollectConsumer(customerSchema, receiveCoordinatorIngest, workerIDs);
+
+    RelationKey diffRelationKey = new RelationKey("public", "adhoc", "diffResult");
+    Difference diff = new Difference(workerConsumerParallelIngest, workerConsumerCoordinatorIngest);
+    DbInsert workerIngest = new DbInsert(diff, diffRelationKey, true);
+
+    workerPlansDiff.put(workerIDs[0], new RootOperator[] {
```

Comment:
Sorry for being dense but I don't quite follow what's going on here; perhaps we could whiteboard in person?